scripts/src/java/org/oppia/android/scripts/coverage/RunCoverage.kt

```diff
@@ -137,16 +141,49 @@ class RunCoverage(
         println("\nComputed Coverage Ratio is: $computedCoverageRatio")
         println("\nGenerated report at: $reportOutputPath\n")
       }
-    } ?: println("No coverage reports generated.")
-  }
 
-  private fun runCoverageForTarget(testTarget: String): CoverageReport {
-    return runBlocking {
-      CoverageRunner(rootDirectory, scriptBgDispatcher, commandExecutor)
-        .runWithCoverageAsync(testTarget.removeSuffix(".kt"))
-        .await()
+      println("COVERAGE ANALYSIS COMPLETED.")
     }
   }
+
+  private fun runCoverageForTarget(testTarget: String): Deferred<CoverageReport> {
+    return CoverageRunner(rootDirectory, scriptBgDispatcher, commandExecutor)
+      .runWithCoverageAsync(testTarget.removeSuffix(".kt"))
+  }
+}
+
+private fun calculateAggregateCoverageReport(
+  coverageReports: List<CoverageReport>
+): CoverageReport {
+  fun aggregateCoverage(coverages: List<Coverage>): Coverage {
+    return if (coverages.contains(Coverage.FULL)) Coverage.FULL
+    else Coverage.NONE
+  }
+
+  val allCoveredLines = coverageReports.flatMap { it.coveredLineList }
+
+  val groupedCoveredLines = allCoveredLines.groupBy { it.lineNumber }
+
+  val aggregatedCoveredLines = groupedCoveredLines.map { (lineNumber, coveredLines) ->
+    CoveredLine.newBuilder()
+      .setLineNumber(lineNumber)
+      .setCoverage(aggregateCoverage(coveredLines.map { it.coverage }))
+      .build()
+  }
+
+  val totalLinesFound = aggregatedCoveredLines.size
+  val totalLinesHit = aggregatedCoveredLines.count { it.coverage == Coverage.FULL }
+
+  val aggregatedTargetList = coverageReports.joinToString(separator = ", ") { it.bazelTestTarget }
+
+  return CoverageReport.newBuilder()
+    .setBazelTestTarget(aggregatedTargetList)
+    .setFilePath(coverageReports.first().filePath)
+    .setFileSha1Hash(coverageReports.first().fileSha1Hash)
```

Comment:
Is it always correct to use the first occurrences for these two properties?

Reply:
It should be and it probably wouldn't matter whether we acquire the first or second or last, as every case for these 2 fields would contain same exact value for the provided file. 

---

So, the **`coverageReports`** will have the list of coverage data (as protos) for each `testTarget` related to a **file**.

**For example:**
The file -> **app/.../StoryActivity.kt** it will produce 2 test files 
- sharedTest/.../StoryActivityTest.kt
- test/.../StoryActivityLocalTest.kt

With their retrieved test targets -> their respective coverage data are collected -> converted to protos -> saved as list of protos for each testTarget in the variable `coverageReports`.

```
val coverageReports = testTargets.map { testTarget ->
    CoverageRunner() ...
}
```

These coverage data will be related to the one single source file.
- sharedTest/.../**StoryActivityTest**.kt would also have filePath to app/.../**StoryActivity.kt**
- test/....../**StoryActivityLocalTest**.kt would also have filePath to app/.../**StoryActivity.kt**

**since both are retrieving coverage data for the same file app/.../StoryActivity.kt**

So, it's just **one single filePath** and **one file's sha hash value** and collecting either of the filePath would result in the same result.

**`first()`** is used for as with most cases it has only one test file resulting in just one coverageReport data list item.

Would this implementation have any concerns?

Reply:
In that case, should we be using ``single`` instead in order to fail if it doesn't have exactly one file (since it should)?

Reply:
Okay, so I wasn't actually checking that case because while producing the coverage report protos we do only collect data that matches the exact target name so there shouldn't be any possibility of multiple file paths or SHA hashes appearing.

**CoverageRunner.kt**
```kotlin
val sfStartIdx = coverageData.indexOfFirst {
  it.startsWith("SF:") && it.substringAfter("SF:").substringAfterLast("/") == extractedFileName
}
```

But it shouldn't hurt to have an extra check and also, earlier I directly acquired the file path and SHA hash values from the coverageReports. But now as suggested and needed I grouped them by file paths and SHA value pairs,

```kotlin
val groupedCoverageReports = coverageReports.groupBy {
  Pair(it.filePath, it.fileSha1Hash)
}

val singleCoverageReport = groupedCoverageReports.entries.single()
```

to confirm that only one value of the filePath and fileSha1Hash values are present for a single file presented to analyse coverage.

So, the entire implementation is tested with this sample data - https://github.com/oppia/oppia-android/pull/5459#issuecomment-2253028530 and adjusting the filePath in one of the items does throw the following error, confirming the check.

![image](https://github.com/user-attachments/assets/d5579106-e5c1-4ab3-aa7b-fdd6df04430d)

Reply:
I think the new flow with groupBy is really clean and feels much more robust. Is it also possible to have an automated test to verify this behavior, too, since you could verify it manually?

Reply:
Ok, I saw this test case request coming and had already kinda thought about it. However, I don't think it would be possible to add the test case for this added exception scenario. 

(I could be wrong though) but,
The reason is I tried replicating the issue by modifying or hardcoding the CoverageReport values instead of acquiring them dynamically - https://github.com/oppia/oppia-android/pull/5459/commits/8a12cf4548796a88c8669ec3ffd23e00a4811b73 (In this commit I modified the filePath for one of the reports to be `AddNums1.kt` to create the exception scenario, but I suspect this scenario might not be possible given our current setup.

```kotlin
val extractedFileName = "${extractTargetName(bazelTestTarget)}.kt"

val sfStartIdx = coverageData.indexOfFirst {
  it.startsWith("SF:") && it.substringAfter("SF:").substringAfterLast("/") == extractedFileName
}
```

We only retrieve the data with the required file (extracted from the bazel test target) name data from the batch of data, and based on this, the specific line of code to handle varied filePaths might never be hit. May be if wrong, I might gain a better understanding after further discussion but as of now I don't think its possible.

But if that's a possible scenario, then may be any hints on when it might occur could be very helpful for me to figure out a way to add tests.

Reply:
As discussed during the meeting, I think the existing tests probably handle this case well enough.